File: КОМПОНЕНТЫ/OPartner/Examples/Delphi/extbl1u.pas
unit ExTbl1u;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, Grids, OpModels, OpShared, OpWrd2k, OpWord;

type
  TForm1 = class(TForm)
    btnUpdate: TButton;
    StringGrid1: TStringGrid;
    Label2: TLabel;
    OpWord1: TOpWord;
    OpEventModel1: TOpEventModel;
    procedure btnUpdateClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure OpEventModel1GetColCount(Sender: TObject;
      var ColCount: Integer);
    procedure OpEventModel1GetColHeaders(Sender: TObject;
      var ColHeaders: Variant);
    procedure OpEventModel1GetData(Sender: TObject; Index, Row: Integer;
      Mode: TOpRetrievalMode; var Size: Integer; var Data: Variant);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

const
  ProjectCaption = 'Populating a Word Table with TDvEventModel';

var
  Doc : TOpWordDocument;
  Table : TOpDocumentTable;
  Headers : Variant;

procedure TForm1.FormCreate(Sender: TObject);
begin
  Caption := ProjectCaption;
  { Turn on Word and create new document }
  OpWord1.Connected := True;
  Doc := OpWord1.Documents.Add;
  Doc.Insert(Caption + #13#10#10#10, True);

  { Create variant array for the table header row }
  Headers := VarArrayCreate([0, 2], varVariant);
  Headers[0] := 'Col 1';
  Headers[1] := 'Col 2';
  Headers[2] := 'Col 3';

  { Create table, hook up with TDvEventModel, and populate }
  Table := Doc.Tables.Add;
  Table.OfficeModel := OpEventModel1;
  OpEventModel1.RowCount := StringGrid1.RowCount;
  Table.PopulateDocTable;
  OpWord1.Visible := True;
end;

procedure TForm1.btnUpdateClick(Sender: TObject);
begin
  { Re-Populate Table }
  Table.RePopulateDocTable;
end;

procedure TForm1.OpEventModel1GetColCount(Sender: TObject;
  var ColCount: Integer);
begin
  { Populating table, need column count }
  ColCount := StringGrid1.ColCount;
end;

procedure TForm1.OpEventModel1GetColHeaders(Sender: TObject;
  var ColHeaders: Variant);
begin
  { Populating table, need variant array of headers }
  ColHeaders := Headers;
end;

procedure TForm1.OpEventModel1GetData(Sender: TObject; Index, Row: Integer;
  Mode: TOpRetrievalMode; var Size: Integer; var Data: Variant);
var
  OleStr : OleVariant;
begin
  { Populating table, need data for specified table cell }
  OleStr := StringGrid1.Cells[Index, Row];
  Data := OleStr;
  Size := Length(OleStr);
end;

procedure TForm1.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  if Assigned(Table) then
    Table.Free;
  if Assigned(Doc) then
    Doc.Free;
  OpWord1.Connected := False;
end;

end.

File: КОМПОНЕНТЫ/OPartner/Examples/CBuildr5/extbl1u.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "ExTbl1u.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "OpModels"
#pragma link "OpShared"
#pragma link "OpWord"
#pragma link "OpWrd2k"
#pragma resource "*.dfm"
TForm1 *Form1;
TOpWordDocument* Doc;
TOpDocumentTable* Tbl;
Variant Headers;

//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
    : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  // Turn on Word and create new document
  OpWord1->Connected = true;
  Doc = OpWord1->NewDocument();
  Doc->Insert(Caption + "/n/n/n", true);

  // Create variant array for the table header row
  int Bounds[2] = {0,2};
  Headers = VarArrayCreate(Bounds, 1, varVariant);
  Headers.PutElement("Col 1", 0);
  Headers.PutElement("Col 2", 1);
  Headers.PutElement("Col 3", 2);

  // Create table, hook up with TDvEventModel, and populate
  Tbl = Doc->Tables->Add();
  Tbl->OfficeModel = OpEventModel1;
  OpEventModel1->RowCount = StringGrid1->RowCount;
  Tbl->PopulateDocTable();
  OpWord1->Visible = true;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnUpdateClick(TObject *Sender)
{
  Tbl->RePopulateDocTable();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OpEventModel1GetColCount(TObject *Sender,
      int &ColCount)
{
  ColCount = StringGrid1->ColCount;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OpEventModel1GetColHeaders(TObject *Sender,
      Variant &ColHeaders)
{
  ColHeaders = Headers;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OpEventModel1GetData(TObject *Sender, int Index,
      int Row, TOpRetrievalMode Mode, int &Size, Variant &Data)
{
  // Populating table, need data for specified table cell
  Variant OleStr = StringGrid1->Cells[Index] [Row];
  Data = OleStr;
  Size = sizeof(OleStr);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormClose(TObject *Sender, TCloseAction &Action)
{
  if (Tbl)
    delete Tbl;
  if (Doc)
    delete Doc;
  OpWord1->Connected = false;
}
//---------------------------------------------------------------------------
